### DISCUSSION TITLE
Kubernetes Helm Chart - apache-airflow-providers-apache-spark provider installation - Help needed

### DISCUSSION THREAD

**Discussion by BJWRD** [apache/airflow#37012] [Q&A] [answered]
Hi All,

I'm attempting to install the apache-airflow-providers-apache-spark provider to my webserver/worker/scheduler pods upon initialisation (see Helm Chart below), so I can then establish successful connectivity to my external Apache Spark K8s app.

    values = [
    <<-EOF
    config:
     core:
      dags_folder: /opt/airflow/dags/repo/ #Required for the scheduler to detect changes within the git-sync repository
      test_connection: enabled
    createUserJob:
      useHelmHooks: false
      applyCustomEnv: false
    migrateDatabaseJob:
      useHelmHooks: false
      applyCustomEnv: false
    executor: "KubernetesExecutor"
    webserverSecretKeySecretName: foo-secret-1
    dags:
      gitSync:
       enabled: true
       repo: ssh://foo
       branch: main
       sshKeySecret: foo-secret-2
    scheduler:
     command: ["sh", "-c"]
     args: 
      - "pip install apache-airflow-providers-apache-spark"
      - "exec airflow webserver" 
    webserver:
     command: ["sh", "-c"]
     args: 
      - "pip install apache-airflow-providers-apache-spark"
      - "exec airflow webserver"
    workers:
     command: ["sh", "-c"]
     args: 
      - "pip install apache-airflow-providers-apache-spark"
      - "exec airflow webserver"
     EOF
  ]

I've been scouring the internet to see how other people have got this to work successfully and have seen a couple of suggestions that an image customisation might be the most suitable option when it comes to installing additional pip packages to the airflow containers? 

Any thoughts/suggestions would be greatly appreciated.

**Comment by potiuk** [comment, marked as answer]
Yes. it is the best approach. https://airflow.apache.org/docs/helm-chart/stable/production-guide.html#extending-and-customizing-airflow-image

**Reply by BJWRD** [reply]
Thanks for the response @potiuk, I think it was probably yourself where I first saw the mention of image customisation for this.

One other question, I've also been deploying Airflow, exec-ing into both the scheduler and webserver pods and then running a `pip install apache-airflow-providers-apache-spark` manually with limited success. The provider installs ok and then after a restart of the respective services I'm able to at least run a DAG which includes the `SparkSubmitOperator`. 

Once the DAG has been run I then receive errors which imply that the ` apache-airflow-providers-apache-spark` is missing  - do you suggest that once again the image customisation is the way forward in order to get the provider baked into Airflow?

**Reply by potiuk** [reply]
Yes. If you decided to use containers, you should go deeper and understand how they work and they work exactly as you described. no changes made to the installation are persistent . This is some super-basic property of containers. This is the main reason why you want to use containers. I strongly recommend to read more about containers (there are plenty tutorials) and getting the hang of it - that will save you - and those who you ask for help - a lot of time.
